src/java.base/share/classes/jdk/internal/ref/CleanerImpl.java

```diff
@@ -138,10 +143,30 @@ public void run() {
                 // Clear the thread locals
                 mlThread.eraseThreadLocals();
             }
+            if (forceCleanup) {
+                try {
+                    synchronized (phantomCleanableList) {
+                        PhantomCleanable<?> next = phantomCleanableList;
+                        do {
+                            next = next.cleanIfNull();
+                        } while (next != phantomCleanableList);
+                    }
+                    synchronized (this) {
+                        cleanupComplete = true;
+                        notify();
+                        while (forceCleanup) {
+                            wait();
+                        }
+                        cleanupComplete = false;
```

Comment:
AFIAU this synchronizes with beforeCheckpoint to verfiy cleanupComplete==true is noticied. An alternative would be to drop this sync, releasing cleaner sooner, and then just reset cleanupComplete in beforeCheckpoint(). But this is not a big deal.

Reply:
This has a point, the waiting is not needed in here. I've moved the synchronization from previous version rather mindlessly.

Reply:
In the current state, at initialization `forceCleanup == cleanupComplete == false`. But after synchronization has been started, `forceCleanup == !cleanupComplete`, always. It seems possible to drop cleanupComplete, right?

Reply:
Right. I also realized that I didn't handle thread termination well; had the thread exited just after setting `forceCleanup`, the checkpoint thread might end up waiting for a notify that never comes. I am not sure if this can happen in practice, because one of the refs in the list is the cleaner itself, but let's be on the safe side.